Fixes: #4034 - Adds Prefix & IPAddress Ordering Tests

DanSheps opened the pull request:
### Fixes: #4034

Adds various tests for Prefix ordering and IPAddress tests, specifically to help address #2562

DanSheps committed:
#4034 - Create tests for prefixes

DanSheps committed:
#4034 - Create tests for addresses

DanSheps committed:
Remove complex ordering for IP addresses

After review complex ordering does not appear to be required

DanSheps committed:
Cleanup Imports

DanSheps committed:
Fix PEP8 errors and document functions

jeremystretch commented on netbox/ipam/tests/test_ordering.py:
```diff
@@ -0,0 +1,290 @@
+    vrfs = None
+
+    def setUp(self):
+        vrfa = VRF(name="VRF A")
```
Should be able to simplify this as:

```python
self.vrfs = (
    VRF(name="VRF A"),
    VRF(name="VRF B"),
    VRF(name="VRF C"),
)
VRF.objects.bulk_create(self.vrfs)
```

DanSheps replied:
Yeah, it should.  I will get on it later today

jeremystretch commented on netbox/ipam/tests/test_ordering.py:
```diff
@@ -0,0 +1,290 @@
+        self.vrfs = (vrfa, vrfb, vrfc)
+
+    # Function to create all prefixes
+    def _create_prefix(self, prefixes):
```
Two issues with this:

1. `prefix` implies a singular instance, but it deals with a list of instances.
2. It's not actually _creating_ the prefixes, just instantiating them.

Something like `prefix_from_args()` would be more accurate. IMO it would be fine to have this deal with a single object, and call it inside list comprehension during creation:

```
Prefix.objects.bulk_create([prefix_from_args(args) for args in prefixes])
```

jeremystretch commented on netbox/ipam/tests/test_ordering.py:
```diff
@@ -0,0 +1,290 @@
+        return (qsprefixes, regprefixes)
+
+    def test_prefix_ordering(self):
+        # Setup Prefixes
```
Let's add docstrings to each test method explaining why each one exists.

jeremystretch commented on netbox/ipam/tests/test_ordering.py:
```diff
@@ -0,0 +1,290 @@
+        self._compare_prefix(Prefix.objects.all(), prefixes)
+
+    def test_prefix_complex_ordering(self):
+        # Setup VRFs
```
Why not just take this approach with the first two tests? Intermixing the ranges makes for a more robust test and obviates the need for a separate method.

DanSheps replied:
We can definitely do this, my thought was this gives us three tests to see what exactly we broke.  I could wrap the three tests into 1 function however.

DanSheps replied:
I have wrapped the first two up into one class, but I think I want to keep this one, as this will help us sort out the nesting of prefix problem in #2562

jeremystretch commented on netbox/ipam/tests/test_ordering.py:
```diff
@@ -0,0 +1,290 @@
+        return addressobjects
+
+    # Function to compare all addresses
+    def _compare_address(self, queryset, addresses):
```
Can we replace this and `_compare_prefix()` above with a single common function?

jeremystretch commented on netbox/ipam/tests/test_ordering.py:
```diff
@@ -0,0 +1,290 @@
+        self.vrfs = (vrfa, vrfb, vrfc)
+
+    # Function to build all addresses
+    def _create_address(self, addresses):
```
Same feedback as `_create_prefix()` above. (Also, can we combine them?)

DanSheps replied:
I will see what I can do.

DanSheps committed:
Update tests

Add docstrings
Consolidate tests

DanSheps committed:
Merge branch '4034-prefix_ordering_tests' of https://github.com/netbox-community/netbox into 4034-prefix_ordering_tests

 Conflicts:
	netbox/ipam/tests/test_ordering.py

jeremystretch commented:
Looks like this has some PEP8 failures.

DanSheps commented:
Yeah, I missed those

On Feb. 11, 2020 13:24, Jeremy Stretch <notifications@github.com> wrote:

Looks like this has some PEP8 failures.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/netbox-community/netbox/pull/4068?email_source=notifications&email_token=ACUJWQFO4QFIFRBOE7REOXDRCL3OBA5CNFSM4KOKCDTKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELNXIKI#issuecomment-584807465>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACUJWQC7IL7TQIFN557M4O3RCL3OBANCNFSM4KOKCDTA>.

DanSheps committed:
Fix PEP8 errors

DanSheps committed:
Fix typo in docstring

jeremystretch commented on netbox/ipam/tests/test_ordering.py:
```diff
@@ -0,0 +1,182 @@
+
+class PrefixOrderingTestCase(OrderingTestBase):
+
+    """
```
Docstrings should go inside their respective methods:

```python
def test_foo(self):
    """
    Docstring goes here
    """

jeremystretch commented on netbox/ipam/tests/test_ordering.py:
```diff
@@ -0,0 +1,182 @@
+    """
+    This is for comparing the complex ordering test case
+    """
+    def _compare_complex(self, queryset, prefixes):
```
It probably makes sense to move this to `OrderingTestBase` with `_compare()` even if it's only used in the one subclass for now.

jeremystretch commented on netbox/ipam/tests/test_ordering.py:
```diff
@@ -0,0 +1,182 @@
+
+        # Setup Prefixes
+        prefixes = (
+            {"status": PrefixStatusChoices.STATUS_CONTAINER, "vrf": None, "family": 4, "prefix": netaddr.IPNetwork('192.168.0.0/16')},
```
Instead of defining all the attrs in dicts first, you can create `Prefix` instances directly in a list, and then pass that list to `Prefix.objects.bulk_create()`. This cuts down on boilerplate somewhat.

```python
prefixes = (
    Prefix(status=PrefixStatusChoices.STATUS_CONTAINER, vrf=None, family=4, prefix=netaddr.IPNetwork('192.168.0.0/16'),
    ...
)
Prefix.objects.bulk_create(prefixes).
```

Plus, you can reference the prefixes by their index on the `prefixes` list even after creation in the database if needed:

```python
prefixes[4].pk
```

DanSheps committed:
Fixes: #4130

Corrects service name for the RQ worker service in docs/installation/upgrading.md
